### PR TITLE
Add support of references/dependencies for Protobuf schema type

### PR DESCRIPTION
**Pull request by libretto**
# What it does
These changes add basic support of references/dependencies for protobuf schema type.
This code contains much common code which allows adding support of references/dependencies to other types.
Limitations:
The compatibility check does not support references/dependencies.
Serialization/Deserialization of protobuf does not support references/dependencies.

References: #195

UPDATE: This PR already supports compatibility check (schema compare).

### CODE REVIEW

**Review by jjaakola-aiven — CHANGES_REQUESTED** [review]
I am getting 500 error with following requests:
```bash
#!/bin/bash

NO_REF='{"schemaType":"PROTOBUF","schema":"syntax = \"proto3\";\n\nmessage NoReference {\n    string name = 1;\n}\n"}'
curl -H "Content-Type: application/json" localhost:8081/subjects/sub1/versions -d "$NO_REF"

WITH_FIRST_REF='{"schemaType":"PROTOBUF","schema":"syntax = \"proto3\";\n\nimport \"NoReference.proto\";\n\nmessage WithReference {\n    string name = 1;\n    NoReference ref = 2;\n}","references":[{"name":"NoReference.proto","subject":"sub1","version":1}]}'
curl -H "Content-Type: application/json" localhost:8081/subjects/sub2/versions -d "$WITH_FIRST_REF"

NO_REF_SECOND='{"schemaType":"PROTOBUF","schema":"syntax = \"proto3\";\n\nmessage NoReferenceTwo {\n    string name = 1;\n}\n"}'
curl -H "Content-Type: application/json" localhost:8081/subjects/sub3/versions -d "$NO_REF_SECOND"

ADD_NEW_REF_IN_SUB2='{"schemaType":"PROTOBUF","schema":"syntax = \"proto3\";\n\nimport \"NoReference.proto\";\nimport \"NoReferenceTwo.proto\";\n\nmessage WithReference {\n    string name = 1;\n    NoReference ref = 2;\n    NoReferenceTwo refTwo = 3;\n}\n","references":[{"name":"NoReference.proto","subject":"sub1","version":1},{"name":"NoReferenceTwo.proto","subject":"sub3","version":1}]}'
curl -H "Content-Type: application/json" localhost:8081/subjects/sub2/versions -d "$ADD_NEW_REF_IN_SUB2"
```

The usage of well known types should be demonstrated in tests and with the REST proxy.

File: karapace/schema_reader.py [hunk 1]
```diff
@@ -448,7 +451,10 @@ def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
                 return
         elif schema_type_parsed == SchemaType.PROTOBUF:
             try:
-                parsed_schema = ValidatedTypedSchema.parse(SchemaType.PROTOBUF, schema_str)
+                resolved_references, resolved_dependencies = self.resolve_schema_references(value)
+                parsed_schema = ValidatedTypedSchema.parse(
+                    SchemaType.PROTOBUF, schema_str, resolved_references, resolved_dependencies
+                )
```
Comment (jjaakola-aiven) [inline karapace/schema_reader.py:457]: When the schema message is a delete there are no references in the data and resolved references and dependencies are `None`.
The parsing will fail as the data in this case will have unresolved dependency.

Register following schemas to subject one and subject two.
```
------------------------------------
Schema to be referenced
------------------------------------
syntax = "proto3";

message NoReference {
  string name = 1;
}

------------------------------------
Schema with reference to subject one schema NoReference.
------------------------------------
syntax = "proto3";

import "NoReference.proto";

message WithReference {
  string name = 1;
  NoReference ref = 2;
}
```

Try to delete the schema from subject two, `DELETE /subjects/subject_two/versions/1`. API returns 200, but deletion fails internally in schema reader, can be seen from logs.

**Review by jjaakola-aiven — CHANGES_REQUESTED** [review]

**Comment by jjaakola-aiven** [comment]
@libretto I will merge this to feature branch in Karapace and I will add some tests and changes on top of this. Please comment on the open questions, especially the need of `.proto` files.

File: karapace/protobuf/google/type/calendar_period.proto [hunk 2]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2021 Google LLC
```
[the remark this answers is not in the document]
Comment (libretto) [reply karapace/protobuf/google/type/calendar_period.proto:1]: We may remove them yet. I build KnownDendencies class by these proto files. While original proto files are unchanged we need not proto files anymore.

File: karapace/protobuf/known_dependency.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        "google/type/interval.proto": [".google.type.Interval"],
+        "google/type/latlng.proto": [".google.type.LatLng"],
+        "google/type/money.proto": [".google.type.Money"],
+        "google/type/month.proto": [".google.type.Month"],
+        "google/type/phone_number.proto": [".google.type.PhoneNumber"],
+        "google/type/postal_address.proto": [".google.type.PostalAddress"],
+        "google/type/quaternion.proto": [".google.type.Quaternion"],
+        "google/type/timeofday.proto": [".google.type.TimeOfDay"],
+        "confluent/meta.proto": [".confluent.Meta"],
+        "confluent/type/decimal.proto": [".confluent.type.Decimal"],
```
[the remark this answers is not in the document]
Comment (libretto) [reply karapace/protobuf/known_dependency.py:120]: KnownDependencise class is built by these proto files. So they may be removed I think.

**Comment by jjaakola-aiven** [comment]
@libretto some of the issues are resolved in PR #473. Please do not update to this PR.

**Comment by libretto** [comment]
> @libretto some of the issues are resolved in PR #473. Please do not update to this PR.

@jjaakola-aiven all things which You resolved was resolved in my code too but i do not push it to GitHub while tests are not successful. 
Now all tests are successful and I push it there. The next step is testing and polishing it.

**Comment by libretto** [comment]
@jjaakola-aiven latest code from https://github.com/instaclustr/karapace/tree/pr2_clear works ok. So we could add Your tests there to see if it works. If something does not work let me fix the bugs.

File: karapace/schema_reader.py (hunk 1 above)
Comment (libretto) [reply karapace/schema_reader.py:458]: fixed

File: karapace/protobuf/schema.py [hunk 4]
```diff
@@ -1,15 +1,21 @@
 # Ported from square/wire:
+# Ported from square/wire:
```
[the remark this answers is not in the document]
Comment (libretto) [reply karapace/protobuf/schema.py]: fixed

File: karapace/protobuf/known_dependency.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (libretto) [reply karapace/protobuf/known_dependency.py]: fixed

File: tests/integration/test_schema_protobuf.py [hunk 5]
```diff
@@ -167,3 +167,180 @@ async def test_protobuf_schema_normalization(registry_async_client: Client, trai
...
+
+    customer_schema = """
+                |syntax = "proto3";
+                |package a1;
+                |import "Place.proto";
+                |message Customer {
+                |        string name = 1;
+                |        int32 code = 2;
+                |        Place place = 3;
+                |}
```
[the remark this answers is not in the document]
Comment (libretto) [reply tests/integration/test_schema_protobuf.py]: ok, i add as reauested to my tests

**Comment by libretto** [comment]
@jjaakola-aiven When I open this PR I notice in the description its limitations and You know that I still working on it. I do not understand why You go to refactor and change code which under my development without any discussion or requests for changes. I see You change the target branch for this PR and Your own branch PR to Main... But Your branch does not contain all my tests and all my code which includes recursive compatibility check which is already released in my code. The question is what I must do? Will You add this functionality by yourself so my branch and this PR are not required anymore so I may close my PR and will continue my next task on top of Your code? Or You will get my latest code and add Your changes to it so You will have all the functionality which I already implemented with Your changes?

**Comment by libretto** [comment]
@jjaakola-aiven i added there more tests from my branch which show bugs in dependencies compatibility algorithm which You implemented by self. And now I try debugging code which You offer there.

**Comment by jjaakola-aiven** [comment]
@libretto Pick the #473 as a base, add missing features, tests, corrections on top of it. That will serve as the feature branch to land the profobuf reference support to Karapace.

**Comment by libretto** [comment]
@jjaakola-aiven please check and merge with Your branch

**Comment by sujayinstaclustr** [comment]
@jjaakola-aiven can you confirm if there is anything that needs to be done, we are waiting for you to approve this PR for merge.

**Comment by libretto** [comment]
@jjaakola-aiven I see there red text with 1 change request but all change requests was resolved weeks ago. I fixed all bugs which we talk about in the code. But this PR was not accepted. Why?

**Comment by RyanSkraba** [comment]
Hello @libretto, I just wanted to keep you in the loop -- I'll be taking a closer look at this PR and #504 in order to get this moving for the next release of Karapace.  Thanks for all the work so far!

I'm catching up a bit on this and the the work being done in #195

**Pull request closed without merging**